### PR TITLE
Self-hosted onboarding first admin wizard

### PR DESCRIPTION
**Pull request by leog**
## What does this PR do?

Implements a wizard to create the first user (admin) as an on-boarding step to the app. The on-boarding will continue to expand in the near future which is why it was implemented this way.

UI:
<kbd>image</kbd>

Can't be used again once at least one user exists in the DB:
<kbd>image</kbd>

Implements #3384 

**Environment**: Staging(main branch) / Production

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?

Remove all users from the DB and go to the login page (not having a session active). It will show the new onboarding wizard. You can also test going to `/auth/setup` having users in the DB to see the second screen above, indicating the step is done (not needed). The API that supports this new functionality also checks if there is no user in the DB to be able to enable creating an admin user through the new onboarding wizard.

### CODE REVIEW

File: apps/web/pages/api/auth/setup.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      identityProvider: IdentityProvider.CAL,
+    },
+  });
+
+  res.status(201).json({ message: "First admin user created successfuly." });
+}
+
+export default defaultHandler({
+  POST: Promise.resolve({ default: defaultResponder(handler) }),
+});
```
Comment (leog) [inline apps/web/pages/api/auth/setup.ts:58]: Added `defaultResponder` to handle zod errors as suggested by @zomars in another PR 👍

File: apps/web/pages/auth/login.tsx [hunk 2]
```diff
@@ -217,6 +218,17 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
     };
   }
 
+  const userCount = await prisma.user.count();
+  if (userCount === 0) {
```
Comment (leog) [inline apps/web/pages/auth/login.tsx:222]: Setup wizard comes up when no user is present.

File: apps/web/pages/auth/setup.tsx [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+const schema = z.object({
+  username: z.string().min(1),
+  email: z.string().email({ message: "Please enter a valid email" }),
+  fullname: z.string(),
+  password: z.string().refine((val) => isPasswordValid(val.trim()), {
+    message:
+      "The password must be a minimum of 7 characters long containing at least one number and have a mixture of uppercase and lowercase letters",
+  }),
+});
```
Comment (leog) [inline apps/web/pages/auth/setup.tsx:24]: `zod` 💓

File: apps/web/pages/auth/setup.tsx [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+const StepDone = () => (
+  <div className="min-h-36 my-6 flex flex-col items-center justify-center">
+    <div className="flex h-[72px] w-[72px] items-center justify-center rounded-full bg-gray-600 dark:bg-white">
+      <CheckIcon className="inline-block h-10 w-10 text-white dark:bg-white dark:text-gray-600" />
+    </div>
+    <div className="max-w-[420px] text-center">
+      <h2 className="mt-6 mb-1 text-lg font-medium dark:text-gray-300">All done!</h2>
+    </div>
+  </div>
+);
```
Comment (leog) [inline apps/web/pages/auth/setup.tsx:35]: Should we use translations?

File: apps/web/pages/auth/setup.tsx [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+  );
+};
+
+export default function Setup(props: inferSSRProps<typeof getServerSideProps>) {
+  const steps = [
+    {
+      title: "Administrator user",
+      description: "Let's create the first administrator user.",
+      content: props.userCount !== 0 ? <StepDone /> : <SetupFormStep1 />,
+      enabled: props.userCount === 0,
```
Comment (leog) [inline apps/web/pages/auth/setup.tsx:265]: This is to check if the wizard should show buttons to navigate through more steps.

File: packages/lib/auth.ts [hunk 6]
```diff
@@ -9,3 +9,19 @@ export async function verifyPassword(password: string, hashedPassword: string) {
   const isValid = await compare(password, hashedPassword);
   return isValid;
 }
+
+export function isPasswordValid(password: string) {
```
Comment (leog) [inline packages/lib/auth.ts:13]: Moved this util from `website` to `lib`, after this PR I will change `website` to use it from `lib` too.

File: packages/ui/Stepper.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import Link from "next/link";
+
+type DefaultStep = {
+  title: string;
+};
+
+function Stepper<T extends DefaultStep>(props: { href: string; step: number; steps: T[] }) {
```
Comment (leog) [inline packages/ui/Stepper.tsx:7]: Moved this component from `console` to `ui`, after this PR I will change`console` to use it from `ui` too.

File: packages/ui/WizardForm.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import Stepper from "./Stepper";
+
+type DefaultStep = {
+  title: string;
+  description: string;
+  content: JSX.Element;
+  enabled?: boolean;
+};
+
+function WizardForm<T extends DefaultStep>(props: { href: string; steps: T[] }) {
```
Comment (leog) [inline packages/ui/WizardForm.tsx:12]: Moved this component from `console` to `ui`, after this PR I will change`console` to use it from `ui` too.

File: apps/web/pages/auth/setup.tsx (hunk 4 above)
Comment (PeerRich) [reply apps/web/pages/auth/setup.tsx:35]: yes

File: apps/web/pages/auth/setup.tsx [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        });
+        if (response.status === 201) {
+          router.replace("/auth/login");
+        } else {
+          router.replace(`/auth/setup`);
+        }
+      }}>
+      <div>
+        <label htmlFor="username" className="sr-only">
+          Username
```
Comment (PeerRich) [inline apps/web/pages/auth/setup.tsx:76]: translation

File: apps/web/pages/auth/setup.tsx [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        </div>
+        {formMethods.formState.errors.username && (
+          <p role="alert" className="mt-1 text-xs text-red-500">
+            {formMethods.formState.errors.username.message}
+          </p>
+        )}
+      </div>
+      <div>
+        <label htmlFor="fullname" className="sr-only">
+          Full name
```
Comment (PeerRich) [inline apps/web/pages/auth/setup.tsx:121]: translation

File: apps/web/pages/auth/setup.tsx [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+        </div>
+        {formMethods.formState.errors.fullname && (
+          <p role="alert" className="mt-1 text-xs text-red-500">
+            {formMethods.formState.errors.fullname.message}
+          </p>
+        )}
+      </div>
+      <div>
+        <label htmlFor="email" className="sr-only">
+          Email address
```
Comment (PeerRich) [inline apps/web/pages/auth/setup.tsx:167]: translation

File: apps/web/pages/auth/setup.tsx [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+        </div>
+        {formMethods.formState.errors.email && (
+          <p role="alert" className="mt-1 text-xs text-red-500">
+            {formMethods.formState.errors.email.message}
+          </p>
+        )}
+      </div>
+      <div>
+        <label htmlFor="password" className="sr-only">
+          Password
```
Comment (PeerRich) [inline apps/web/pages/auth/setup.tsx:213]: translation

**Review by PeerRich — APPROVED** [review]
great stuff! i imagine we will extend this with more admin setup stuff like enabling/disabling apps, etc?

**Comment by leog** [comment]
> great stuff! i imagine we will extend this with more admin setup stuff like enabling/disabling apps, etc?

Exactly, apps and custom product name were the suggested additions to this wizard. Feel free to suggest more.

**Comment by zomars** [comment]
@leog can you reopen the PR? This got auto-merged before I was able to review.

**Comment by zomars** [comment]
Reopened in #3486

**Comment by PeerRich** [comment]
oops sorry for that

**Pull request merged**